src/patch/apply.js

```diff
@@ -42,7 +40,61 @@ export function applyPatch(source, uniDiff, options = {}) {
             return false;
           }
         }
+        toPos++;
       }
+    }
+
+    return true;
+  }
+
+ // Search best fit offsets for each hunk based on the previous ones
+ for (let i = 0; i < hunks.length; i++) {
+    let hunk = hunks[i],
+        outOfLimits = 0,
+        localOffset = 0,
+        minLine = 0,
+        toPos = offset + hunk.oldStart - 1;
+
+    for (;;) {
+      if (toPos + localOffset + hunk.oldLines <= lines.length) {
+        if (hunkFits(hunk, toPos + localOffset)) {
+          hunk.offset = offset += localOffset;
+          break;
+        }
+      } else {
+        outOfLimits++;
+      }
+
+      // If trying to fit hunk outside both limits, return error
+      if (outOfLimits === 2) {
+        return false;
+      }
+
+      outOfLimits = 0;
+      localOffset++;
+
+      if (minLine <= toPos - localOffset) {
```

Comment:
Need more comments to describe what is going on in each phase here. It seems like this is scanning both backwards and forwards looking for a match, providing equal weighting to both directions (vs. scanning backwards, then forwards which could bias one outcome) but I'm really not clear. Ideally this would have a high level overview in comments and then links to any articles or similar resources that discuss the underlying algorithm (if this came from something like that)